cooked-validators/src/Cooked/Tx/Constraints/Type.hs

```diff
@@ -48,281 +48,230 @@ import Type.Reflection
-    lens
-      (\output -> Pl.Address (Pl.ScriptCredential . Pl.validatorHash $ output ^. scriptOutputDatumValidatorL) (output ^. scriptOutputDatumCredentialL))
-      (\(ScriptOutputWithDatum _ _ datum value) (validator, mStCred) -> ScriptOutputWithDatum validator mStCred datum value)
-  sOutOutputDatumL =
-    lens
-      (\output -> Pl.OutputDatum (Pl.Datum . Pl.toBuiltinData $ (output ^. scriptOutputDatumDatumL)))
-      (flip (scriptOutputDatumDatumL .~))
-
--- TODO
-scriptOutputWithDatumFromGeneric :: GenericOutput -> Maybe (ScriptOutputWithDatum a)
-scriptOutputWithDatumFromGeneric = undefined
-
--- ** outputs with datum hash belonging to scripts
-
-data ScriptOutputWithDatumHash a where
-  ScriptOutputWithDatumHash ::
-    Show (Pl.DatumType a) =>
-    Pl.TypedValidator a ->
-    Maybe Pl.StakingCredential ->
-    Pl.DatumType a ->
-    Pl.Value ->
-    ScriptOutputWithDatumHash a
-
-deriving instance (Show (ScriptOutputWithDatumHash a))
-
-instance IsOutput (ScriptOutputWithDatumHash a) where
-  type OwnerType (ScriptOutputWithDatumHash a) = Pl.TypedValidator a
-  type DatumType (ScriptOutputWithDatumHash a) = Pl.DatumType a
-  sOutOwnerL = undefined
-  sOutDatumL = undefined
-  sOutValueL = undefined
-  sOutAddressL = undefined
-
-  -- TODO Remember that an 'OutputDatumHash ...' should be targeted by the lens
-  sOutOutputDatumL = undefined
+data ConcreteOutput ownerType datumType where
+  ConcreteOutput ::
+    { concreteOutputOwner :: ownerType,
+      concreteOutputStakingCredential :: Maybe Pl.StakingCredential,
+      concreteOutputValue :: Pl.Value,
+      concreteOutputDatum :: datumType
+      -- concreteOutputReferenceScript :: Maybe Pl.ScriptHash -- TODO for when we introduce reference scripts
+    } ->
+    ConcreteOutput ownerType datumType
+
+deriving instance (Show ownerType, Show datumType) => Show (ConcreteOutput ownerType datumType)
+
+instance
+  ( Show ownerType,
+    ToCredential ownerType,
+    Show datumType,
+    ToOutputDatum datumType
+  ) =>
+  IsOutput (ConcreteOutput ownerType datumType)
+  where
+  type OwnerType (ConcreteOutput ownerType datumType) = ownerType
+  type DatumType (ConcreteOutput ownerType datumType) = datumType
+  outputOwnerL = lens concreteOutputOwner (\out owner -> out {concreteOutputOwner = owner})
+  outputStakingCredentialL = lens concreteOutputStakingCredential (\out mStCred -> out {concreteOutputStakingCredential = mStCred})
+  outputDatumL = lens concreteOutputDatum (\out datum -> out {concreteOutputDatum = datum})
+  outputValueL = lens concreteOutputValue (\out value -> out {concreteOutputValue = value})
+  outputAddress out =
+    Pl.Address
+      (toCredential $ concreteOutputOwner out)
+      (concreteOutputStakingCredential out)
+  outputOutputDatum = toOutputDatum . concreteOutputDatum
+
+-- ** A few special concrete outputs
+
+type PKOutput = ConcreteOutput Pl.PubKeyHash ()
+
+type ScriptOutputWithInlineDatum a = ConcreteOutput (Pl.TypedValidator a) (Pl.DatumType a)
```

Comment:
I might be out of context here, but how do we represent non-inline datums?

I sure don't know the design target, but would `type ScriptOutput a = ConcreteOutput (Pl.TypedValidator a) (TxSkelOutDatum a)` make sense as well?